[LTC] Upstream class BackendDevice

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68027

Summary:
This commit upstreams class BackendDevice to the master, which is a backend
specific representation of the actual hardware, for instances, CPU, GPU, or
TPU.

This concept is important for backend like XLA where it needs to tell the
actual hardware type from the c10::DeviceType::Lazy virtual device during
both IR constructions and lowerings.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D32261838](https://our.internmc.facebook.com/intern/diff/D32261838)

Commit:
[LTC] Upstream class BackendDevice

Summary:
This commit upstreams class BackendDevice to the master, which is a backend
specific representation of the actual hardware, for instances, CPU, GPU, or
TPU.

This concept is important for backend like XLA where it needs to tell the
actual hardware type from the c10::DeviceType::Lazy virtual device during
both IR constructions and lowerings.

Test Plan:
WIP.

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32261838).

alanwaketan commented on torch/csrc/lazy/backend/backend_device.cpp:
```diff
@@ -0,0 +1,41 @@
+  : type_(std::move(type)), ordinal_(ordinal) {}
+
+BackendDevice::BackendDevice(const std::string& device_spec)
+  : type_(std::make_shared<BackendDeviceType>()) {}
```
Here is the fix, @wconstab.

Commit:
Update on "[WIP][LTC] Upstream class BackendDevice"


Summary:
This commit upstreams class BackendDevice to the master, which is a backend
specific representation of the actual hardware, for instances, CPU, GPU, or
TPU.

This concept is important for backend like XLA where it needs to tell the
actual hardware type from the c10::DeviceType::Lazy virtual device during
both IR constructions and lowerings.

Test Plan:
WIP.

Differential Revision: [D32261838](https://our.internmc.facebook.com/intern/diff/D32261838)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32261838).

alanwaketan commented:
This pull request is ready for review.

wconstab commented on torch/csrc/lazy/backend/backend_device.h:
```diff
@@ -0,0 +1,49 @@
+  std::string toString() const;
+
+  // The string -> Device conversion should be handled by the backend interface.
+  C10_DEPRECATED explicit BackendDevice(const std::string& device_spec);
```
am I supposed to delete this right away, once I merge my changes for backend interface? or we need to keep it around for a while longer?

alanwaketan replied:
Good question. I think you can just delete this ctor during or after your merge PR.

JackCaoG replied:
yea, backend can probably just add this if they want to(pt/xla will mostly add it on the extended Device class) .

wconstab commented on torch/csrc/lazy/backend/backend_device.h:
```diff
@@ -0,0 +1,49 @@
+  int8_t type {0};
+
+  virtual ~BackendDeviceType() = default;
+  virtual std::string toString() const { return "Unknown"; }
```
did you consider tradeoffs of making this a pure-interface class vs having implementation that returns 'Unknown'?

alanwaketan replied:
It's probably just a policy. Do we want to enforce all backends to implement this interface? I think backends should implement this but I don't know if we want to force that. Things can still work even if they don't, for instance, TS backend. Just for UAPI that query the actual hardware type won't work. So, I'm leaning towards to recommended but not required.

wconstab replied:
yea that makes sense.  I am fine with this.

Commit:
Update on "[LTC] Upstream class BackendDevice"


Summary:
This commit upstreams class BackendDevice to the master, which is a backend
specific representation of the actual hardware, for instances, CPU, GPU, or
TPU.

This concept is important for backend like XLA where it needs to tell the
actual hardware type from the c10::DeviceType::Lazy virtual device during
both IR constructions and lowerings.

Test Plan:
./build/bin/test_lazy --gtest_filter=BackendDeviceTest.*

Differential Revision: [D32261838](https://our.internmc.facebook.com/intern/diff/D32261838)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32261838).

JackCaoG reviewed (approved):
Thanks!

alanwaketan commented:
Thanks for approving this pull request, @wconstab and @JackCaoG. I will wait for the CI signals before landing.
